atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraServiceTest.java

```diff
@@ -241,34 +254,226 @@ public void selectsHostsWithUnknownDatacenterMappingIfAllKnownDatacentersTried()
+                    .isPresent()
+                    .get()
+                    .satisfies(server -> assertThat(desired).contains(server));
+        }
+    }
+
+    @Test
+    public void testSingleServerGetCassandraServers() throws Exception {
+        Set<CassandraServer> allHosts = ImmutableSet.of(SERVER_1);
+        CassandraService cassandra = clientPoolWithServers(allHosts);
+        TokenRange tokens = new TokenRange();
+        tokens.setEndpoints(
+                allHosts.stream().map(CassandraServer::cassandraHostName).collect(Collectors.toList()));
+        tokens.setEndpoint_details(ImmutableList.of(endpointDetails(HOSTNAME_1, DC_1, RACK_1)));
+        List<TokenRange> tokenRanges = ImmutableList.of(tokens);
+        ImmutableSet<CassandraServer> servers = cassandra.getCassandraServers(tokenRanges);
+        assertThat(servers).hasSize(1).first().satisfies(cs -> {
+            assertThat(cs.cassandraHostName()).isEqualTo(HOSTNAME_1);
+            assertThat(cs.reachableProxyIps()).hasSize(1);
+        });
+    }
+
+    @Test
+    public void testMultiServerGetCassandraServers() throws Exception {
+        Set<CassandraServer> allHosts = ImmutableSet.of(SERVER_1, SERVER_2, SERVER_3);
+        CassandraService cassandra = clientPoolWithServers(allHosts);
+        List<TokenRange> tokenRanges = generateTokenRanges(3, 1, allHosts);
+        assertThat(cassandra.getCassandraServers(tokenRanges)).hasSize(3);
+    }
+
+    @Test
+    public void testComplexMultiDatacenterMultiRackServerGetCassandraServers() throws Exception {
+        int replicas = 3;
+        int partitions = 5;
+        Set<CassandraServer> nodes = generateNodes(replicas, partitions);
+        CassandraService cassandra = clientPoolWithServers(nodes);
+        List<TokenRange> tokenRanges = generateTokenRanges(replicas, partitions, nodes);
+
+        assertThat(cassandra.getCassandraServers(tokenRanges)).hasSize(nodes.size());
+        // spot check some subsets of ranges
+        assertThat(cassandra.getCassandraServers(tokenRanges.subList(0, 1))).hasSize(replicas);
+        assertThat(cassandra.getCassandraServers(tokenRanges.subList(0, 2))).hasSize(2 * replicas);
+        assertThat(cassandra.getCassandraServers(tokenRanges.subList(0, 3))).hasSize(3 * replicas);
+        assertThat(cassandra.getCassandraServers(tokenRanges.subList(1, 3))).hasSize(2 * replicas);
+        assertThat(cassandra.getCassandraServers(tokenRanges.subList(2, 4))).hasSize(2 * replicas);
+        assertThat(cassandra.getCassandraServers(tokenRanges.subList(3, 5))).hasSize(2 * replicas);
+        assertThat(cassandra.getCassandraServers(tokenRanges.subList(4, 5))).hasSize(replicas);
+
+        // exhaustively check all ranges
+        for (int i = 0; i < partitions; i++) {
+            for (int j = i + 1; j <= partitions; j++) {
+                for (int k = 1; k <= replicas; k++) {
+                    List<TokenRange> ranges = tokenRanges.subList(i, j);
```

Comment:
sorry why do we have the k loop?

Reply:
k is replication factor to simulate larger cluster, will adjust to make this clearer